### PR TITLE
Update some pivot QP code to use MLv2 (bug fix)

### LINKED ISSUE #35025: Pivot viz fails on a nested question
### Describe the bug

Pivot viz works while editing a question based on a nested question, but it fails after the question is saved.

### To Reproduce

1. New GUI question: select all records from Products
2. Save it as Question1
3. New GUI question: select Question1; Aggregate: Count of Rows, Sum of Total; Group by Category, CreatedAt: Month
4. Select Pivot table for the visualization
5. The viz shows up correctly
6. Save it as Question2
7. (!) Re-open Question2: the viz fails with _There was a problem with your question / Error running query_

### Expected behavior

It should work

### Logs

```
2023-10-24 18:19:22,516 ERROR middleware.catch-exceptions :: Error processing query: Error running query                                                                                                    [479/1701]
{:json_query
 {:database 1,
  :type :query,
  :query
  {:aggregation [[:count] [:sum [:field "PRICE" {:base-type :type/Float}]]],
   :breakout ([:expression "pivot-grouping"]),
   :source-table "card__1",
   :expressions {:pivot-grouping [:abs -9223372036854775805]}},
  :constraints {:max-results 10000, :max-results-bare-rows 2000},
  :middleware {:js-int-to-string? true, :ignore-cached-results? false, :add-default-userland-constraints? true},
  :async? true},
 :native
 {:query
  "SELECT \"source\".\"pivot-grouping\" AS \"pivot-grouping\", COUNT(*) AS \"count\", SUM(\"source\".\"PRICE\") AS \"sum\" FROM (SELECT \"source\".\"PRICE\" AS \"PRICE\", ABS(-9223372036854775805) AS \"pivot-groupi
ng\" FROM (SELECT \"PUBLIC\".\"PRODUCTS\".\"ID\" AS \"ID\", \"PUBLIC\".\"PRODUCTS\".\"EAN\" AS \"EAN\", \"PUBLIC\".\"PRODUCTS\".\"TITLE\" AS \"TITLE\", \"PUBLIC\".\"PRODUCTS\".\"CATEGORY\" AS \"CATEGORY\", \"PUBLIC
\".\"PRODUCTS\".\"VENDOR\" AS \"VENDOR\", \"PUBLIC\".\"PRODUCTS\".\"PRICE\" AS \"PRICE\", \"PUBLIC\".\"PRODUCTS\".\"RATING\" AS \"RATING\", \"PUBLIC\".\"PRODUCTS\".\"CREATED_AT\" AS \"CREATED_AT\" FROM \"PUBLIC\".\
"PRODUCTS\") AS \"source\") AS \"source\" GROUP BY \"source\".\"pivot-grouping\" ORDER BY \"source\".\"pivot-grouping\" ASC",
  :params nil},                                                                                                                                                                                             [461/1701]
 :status :failed,
 :class java.lang.IndexOutOfBoundsException,
 :stacktrace
 ["clojure.lang.PersistentVector.arrayFor(PersistentVector.java:158)"
  "clojure.lang.PersistentVector.nth(PersistentVector.java:162)"
  "clojure.lang.RT.nth(RT.java:896)"
  "--> query_processor.pivot$generate_queries$iter__94005__94009$fn__94010$iter__94027__94031$fn__94032$fn__94033.invoke(pivot.clj:129)"
"query_processor.pivot$generate_queries$iter__94005__94009$fn__94010$iter__94027__94031$fn__94032.invoke(pivot.clj:128)"

```
### Information about your Metabase installation

```JSON
1.47.5
```

### Severity

Seems severe.

### Additional context

Works when there is no nested query.
Works with 1.47.4.

### PR DESCRIPTION
**Pull request by camsaul**
Fixes #35025

Update some pivot QP code to use MLv2. 

Test is roughly based on #35076

### CODE REVIEW

**Review by metamben — APPROVED** [review]
👍

File: src/metabase/query_processor/pivot.clj [hunk 1]
```diff
@@ -204,25 +208,47 @@
...
-        breakout          (mapv qp.util/field-ref->key
-                                (-> query :query :breakout))
-        index-in-breakout (fn [field-ref]
-                            (.indexOf ^clojure.lang.PersistentVector breakout (qp.util/field-ref->key field-ref)))
-        pivot-rows        (mapv index-in-breakout (:rows column-split))
-        pivot-cols        (mapv index-in-breakout (:columns column-split))]
+  [query        :- [:map
+                    [:database ::lib.schema.id/database]]
+   viz-settings :- [:maybe :map]]
+  (let [index-in-breakouts (delay
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/query_processor/pivot.clj]: good call. fixed

File: src/metabase/lib/equality.cljc [hunk 2]
```diff
@@ -321,7 +323,7 @@
 
   ([query        :- ::lib.schema/query
     stage-number :- :int
-    legacy-ref   :- some?
+    legacy-ref   :- :some #_mbql.s/Reference
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/lib/equality.cljc]: removed

File: src/metabase/lib/equality.cljc [hunk 3]
```diff
@@ -265,9 +265,11 @@
        (when (and query (number? ref-id))
          (when-let [no-id-columns (not-empty (remove :id columns))]
            (when-let [resolved (resolve-field-id query stage-number ref-id)]
-             (find-matching-column (assoc a-ref 2 (or (:lib/desired-column-alias resolved)
-                                                      (:name resolved)))
-                                   no-id-columns opts)))))))
+             (find-matching-column (-> (assoc a-ref 2 (or (:lib/desired-column-alias resolved)
+                                                          (:name resolved)))
+                                       (lib.options/update-options (partial merge {:base-type :type/*})))
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/lib/equality.cljc:274]: good call. fixed

**Pull request merged**